google/cloud/security/scanner/scanners/iap_scanner.py

```diff
@@ -0,0 +1,443 @@
+                        allowed_sources.update(
+                            firewall_rule.source_ranges or [])
+                        allowed_sources.update(
+                            firewall_rule.source_tags or [])
+                        continue
+            for firewall_rule in relevant_rules_by_priority[priority]:
+                for denied in firewall_rule.denied or []:
+                    if firewall_entry_applies(denied):
+                        allowed_sources.difference_update(
+                            firewall_rule.source_ranges or [])
+                        allowed_sources.difference_update(
+                            firewall_rule.source_tags or [])
+        return allowed_sources
+
+    def tags_for_instance_group(self, instance_group):
+        """Which instance tags are used for an instance group?
+
+        Includes tags used by instances in the group and, for managed
+        groups, tags in the group's template.
+
+        Args:
+            instance_group (InstanceGroup): the group to query tags for
+
+        Returns:
+            set: tags
+        """
+        tags = set()
+
+        # Get tags from actual instances.
+        for instance_url in instance_group.instance_urls:
+            instance = self.find_instance_by_url(instance_url)
+            if not instance:
+                continue
+            tags.update(instance.tags)
+
+        # If it's a managed instance group, also get tags from the
+        # instance template.
+        instance_template = self.instance_templates_by_group_key.get(
+            instance_group.key)
+        if instance_template:
+            template_tags = instance_template.properties.get('tags', {})
+            tags.update(template_tags.get('items', []))
+
+        return tags
+
+    def make_iap_resource(self, backend_service):
+        """Get an IapResource for a service.
+
+        Args:
+            backend_service (BackendService): service to create a resource for
+
+        Returns:
+            IapResource: the corresponding resource
+        """
+        alternate_services = set()
+        direct_access_sources = set()
+        for backend in backend_service.backends:
+            instance_group = self.find_instance_group_by_url(
+                backend.get('group'))
+            if not instance_group:
+                continue
+
+            network_port = self.instance_group_network_port(
+                backend_service, instance_group)
+
+            direct_access_sources.update(
+                self.firewall_allowed_sources(network_port, None))
+            tags = self.tags_for_instance_group(instance_group)
+            for tag in tags:
+                direct_access_sources.update(
+                    self.firewall_allowed_sources(
+                        network_port, tag))
+
+        # Don't count the load balancer as a direct access source.
+        direct_access_sources.discard('130.211.0.0/22')
```

Comment:
In theory they can change, and in fact they recently did, but it's a breaking change for customers so it's not something that should happen often, and when it does we give plenty of warning.  Added a comment explaining this.